07_memory/07_kernel_ver/krnl_mem_test.c

```diff
@@ -0,0 +1,197 @@
+
+struct mem_live_time {
+	uint64_t alloc_time;
+	uint64_t free_time;
+};
+
+enum mem_func_name {
+	KMALLOC = 0,
+	KZALLOC,
+	VMALLOC,
+	GET_FREE_PAGES,
+	NUM_FN
+};
+
+enum test_case {
+	TWO_POW_X = 0,
+	TWO_POW_X_PLUS_ONE,
+	NUM_TC
+};
+
+
+int get_time(uint8_t order, enum test_case tc_num,
+		struct mem_live_time *t_mem_live, enum mem_func_name func_name)
+{
+	uint64_t t_start = 0;
+	uint64_t t_stop = 0;
+	char *mem_ptr = NULL;
+	uint64_t size;
+
+	size = (tc_num == TWO_POW_X)?POW2(order):(POW2(order) + 1);
+
+	t_start = ktime_get_ns();
+
+	switch (func_name) {
+
+	case KMALLOC:
+		mem_ptr = kmalloc(size * sizeof(*mem_ptr), GFP_KERNEL);
+		break;
+
+	case KZALLOC:
+		mem_ptr = kzalloc(size * sizeof(*mem_ptr), GFP_KERNEL);
+		break;
+
+	case VMALLOC:
+		mem_ptr = vmalloc(size * sizeof(*mem_ptr));
+		break;
+
+	case GET_FREE_PAGES:
+		if (order > 0)
+			mem_ptr = (void *)__get_free_pages(GFP_KERNEL, order);
+		break;
+
+	default:
+		mem_ptr = NULL;
+		break;
+	}
+
+	t_stop = ktime_get_ns();
+
+	if (mem_ptr == NULL)
+		return 1;
+
+	t_mem_live->alloc_time = t_stop - t_start;
+
+	t_start = ktime_get_ns();
+
+	switch (func_name) {
+
+	case KMALLOC:
+	case KZALLOC:
+		kfree(mem_ptr);
+		break;
+
+	case VMALLOC:
+		vfree(mem_ptr);
+		break;
+
+	case GET_FREE_PAGES:
+		if (order > 0)
+			free_pages((unsigned long)mem_ptr, order);
+		break;
+
+	default:
+		mem_ptr = NULL;
+		break;
+	}
+
+	t_stop = ktime_get_ns();
+
+	t_mem_live->free_time = t_stop - t_start;
+
+	return 0;
+}
+
+void run_test_mem(enum test_case tc_num, enum mem_func_name func_name)
+{
+	struct mem_live_time t_current = {0};
+	uint64_t size = 0;
+	uint64_t t_min_alloc = U64_MAX;
+	uint64_t t_max_alloc = 0;
+	uint64_t t_min_free = U64_MAX;
+	uint64_t t_max_free = 0;
+	uint64_t sum_alloc = 0;
+	uint64_t sum_free = 0;
+	int i = 0;
+	int j = 0;
+
+	pr_info("====== Run memory test allocation! ============================\n");
+	pr_info("|\n");
+	pr_info("| function call: %s\n", mem_func_name[func_name]);
+	pr_info("| test case:     %s\n", (tc_num)?"2^x + 1":"2^x");
+	pr_info("|\n");
+	pr_info("===============================================================\n");
+
+
+	for (i = 0; i < MAX_SIZE; i++) {
+
+		pr_info("| x: %d", i);
+
+		t_min_alloc = U64_MAX;
+		t_max_alloc = 0;
+		t_min_free = U64_MAX;
+		t_max_free = 0;
+
+		for (j = 0; j < LOOP_SIZE; j++) {
+
+			if (get_time(i, tc_num, &t_current, func_name)) {
+				pr_info("| ### Memory cannot be allocated!\n");
+				return;
+			}
+
+			t_min_alloc = MIN(t_current.alloc_time, t_min_alloc);
+			t_max_alloc = MAX(t_current.alloc_time, t_max_alloc);
+			t_min_free = MIN(t_current.free_time, t_min_free);
+			t_max_free = MAX(t_current.free_time, t_max_free);
+
+		}
+
+		pr_info("| alloc time:  min: %llu(ns)"
+				" average: %llu(ns)"
+				" max: %llu(ns)\t"
+				"| free time:  min: %llu(ns)"
+				" average: %llu(ns)"
+				" max: %llu(ns)"
+				, t_min_alloc, (t_max_alloc + t_min_alloc)/2, t_max_alloc,
```

Comment:
Average value is always equal to (max+min)/2 for arrays this size less than 3.
For other series we need calculate it as
sum(values)/quantity(values)

Reply:
The results will be equals, but I think my solution has better performance, because we don't need additional sum variable and operations with it. 
Example. {1 3 5 7 9} Average is: 5 = (1+3+5+7+9)/5 = (1+9)/2.
If we want to calculate average on different allocation function, yes I guess we need to use sum as mentioned above.
Perhaps I incorrect interpreted your comment?

Reply:
What is the average value for array {1, 2, 3, 40} ?

Reply:
11,5.
Yes, you are right absolutely. That's may mistake. 
Fixed.